File: codewars/src/com/juhezi/codewars/kyu6/GridPath.java
package com.juhezi.codewars.kyu6;

import java.math.BigInteger;

/**
 * URL: https://www.codewars.com/kata/paths-in-the-grid/train/java
 * 最简单的动态规划
 */
public class GridPath {

    public static BigInteger numberOfRoutes(int m, int n) {
        BigInteger[][] array = new BigInteger[m + 1][n + 1];
        for (int i = 0; i <= m; i++)
            array[i][0] = BigInteger.ONE;
        for (int j = 0; j <= n; j++) {
            array[0][j] = BigInteger.ONE;
        }
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                array[i][j] = array[i - 1][j].add(array[i][j - 1]);
            }
        }
        return array[m][n];
    }

    public static void main(String[] args) {
        System.out.println(numberOfRoutes(3,4));
    }

}

File: codewars/src/com/juhezi/codewars/kyu6/GridPath.clj
(ns com.juhezi.codewars.kyu6.GridPath)

; URL: https://www.codewars.com/kata/paths-in-the-grid/train/java
; 最简单的动态规划
; 首先要定义一个数组

;递归，这个速度太慢

;(defn number-of-routes [m n]
;  (if (= m 0)
;    1
;    (if (= n 0)
;      1
;      (+ (number-of-routes m (- n 1))
;         (number-of-routes (- m 1) n)))))

(defn fact [s e]
  (reduce * (range (biginteger s) (biginteger (inc e))))
  )
(defn number-of-routes [m n]
  (/ (fact (inc n) (+ n m)) (fact 1 m)))
